servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpRequestTracker.java

```diff
@@ -0,0 +1,108 @@
+import io.servicetalk.loadbalancer.RequestTracker;
+import io.servicetalk.transport.api.ConnectionInfo;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+final class HttpRequestTracker {
+
+    private HttpRequestTracker() {
+        // no instances
+    }
+
+    static FilterableStreamingHttpConnection observe(
+            Function<HttpResponseMetaData, ErrorClass> peerResponseErrorClassifier,
+            Function<Throwable, ErrorClass> errorClassFunction,
+            RequestTracker requestTracker, FilterableStreamingHttpConnection connection) {
+        HttpLifecycleObserverRequesterFilter filter = new HttpLifecycleObserverRequesterFilter(
+                new Observer(peerResponseErrorClassifier, errorClassFunction, requestTracker));
+        return filter.create(connection);
+    }
+
+    private static class Observer implements HttpLifecycleObserver {
+
+        private final Function<HttpResponseMetaData, ErrorClass> peerResponseErrorClassifier;
+        private final Function<Throwable, ErrorClass> errorClassFunction;
+        private final RequestTracker tracker;
+
+        Observer(final Function<HttpResponseMetaData, ErrorClass> peerResponseErrorClassifier,
+                 final Function<Throwable, ErrorClass> errorClassFunction, final RequestTracker tracker) {
+            this.peerResponseErrorClassifier = peerResponseErrorClassifier;
+            this.errorClassFunction = errorClassFunction;
+            this.tracker = tracker;
+        }
+
+        @Override
+        public HttpExchangeObserver onNewExchange() {
+            return new RequestTrackerExchangeObserver();
+        }
+
+        private final class RequestTrackerExchangeObserver implements HttpLifecycleObserver.HttpExchangeObserver {
+
+            // TODO: cleanup.
+            private final AtomicLong startTime = new AtomicLong(Long.MIN_VALUE);
+
+            @Override
+            public void onConnectionSelected(ConnectionInfo info) {
+                // noop
+            }
+
+            @Override
+            public HttpLifecycleObserver.HttpRequestObserver onRequest(HttpRequestMetaData requestMetaData) {
+                startTime.set(tracker.beforeRequestStart());
+                return NoopHttpLifecycleObserver.NoopHttpRequestObserver.INSTANCE;
+            }
+
+            @Override
+            public HttpLifecycleObserver.HttpResponseObserver onResponse(HttpResponseMetaData responseMetaData) {
+                ErrorClass error = peerResponseErrorClassifier.apply(responseMetaData);
+                if (error != null) {
+                    final long startTime = finish();
+                    if (checkOnce(startTime)) {
+                        tracker.onRequestError(startTime, error);
+                    }
+                }
+                return NoopHttpLifecycleObserver.NoopHttpResponseObserver.INSTANCE;
+            }
+
+            @Override
+            public void onResponseError(Throwable cause) {
```

Comment:
`onResponseError`/`onResponseCancel` methods are triggered on `HttpExchangeObserver` only before response meta-data arrive. After that, you need the same methods on `HttpResponseObserver`. To simplify everything, consider:
```
private final class RequestTrackerExchangeObserver implements HttpLifecycleObserver.HttpExchangeObserver, HttpLifecycleObserver.HttpResponseObserver {
```

Reply:
Ah, thank you.